### ISSUE TITLE
nn.Linear backward seems not thread safe

### ISSUE THREAD

**Issue by hym1120** [torch/nn#1202]
Hi,

If I run below code , I got "inf" as gradient.
If I comment out the "setnumthreads" line, it runs fine.

Please advice.
Yaming

require 'torch'
require 'nn'
dtype = 'torch.FloatTensor'
--------------------------
torch.setnumthreads(8)
--------------------------
model = nn.Linear(5,1):type(dtype)
params, grad_params = model:getParameters()
x = torch.rand(2, 5):type(dtype)
y = torch.rand(2, 1):type(dtype)
scores = model:forward(x)
model:backward(x, scores)
print (grad_params)

**Comment by hym1120** [comment]
I am running on linux redhat
version is
2.6.32-573.35.2.el6.x86_64
